Implement bottom tabs and main menu

PIG208 opened the pull request:
This is stacked on top of some commits cherry-picked from #1041:
* action_sheet [nfc]: Use a generalized name for the cancel button
* action_sheet [nfc]: Extract ActionSheetMenuItemButton

See also: [CZO discussion](https://chat.zulip.org/#narrow/channel/48-mobile/topic/Buttons.20on.20the.20bottom.20tabs.20and.20main.20menu)

<details>
<summary>Screenshots</summary>

| light | dark |
| - | - |
|  ![1000015700](https://github.com/user-attachments/assets/2025dbbb-b4b5-4069-b7c5-0881e853297e) | ![1000015702](https://github.com/user-attachments/assets/7fedf362-ec97-4e05-bcd6-93482a8e3b02) |
| ![1000015701](https://github.com/user-attachments/assets/e0473062-3e27-4058-b967-1713dd2e72cc) | ![1000015703](https://github.com/user-attachments/assets/c94976ba-9aff-4dc6-bf6a-e8d663ad59a6) |

</details>

Fixes: #1035

gnprice commented:
(I turned the link to #1035 into the kind GitHub recognizes specially; that's helpful for example because it makes it appear in the issue's row on the [project board]([Bottom tabs and main menu](https://github.com/zulip/zulip-flutter/issues/1035)	https://github.com/zulip/zulip-flutter/issues/1035). I realize the PR isn't ready to be merged as our solution to the issue, but that's expressed by its draft status; and I believe it is intended to be the PR that fixes the issue.)

gnprice commented on lib/widgets/home.dart:
```diff
@@ -0,0 +1,461 @@
+
+  final bool selected;
+
+  @override
+  // TODO: implement icon
+  IconData get icon => ZulipIcons.attach_file;
+
+  @override
+  String label(ZulipLocalizations zulipLocalizations) {
+    return 'Attach file';
+  }
+
+  @override
+  void onPressed() {
+    // TODO: implement onPressed
+  }
```
Oh and I guess there'll also be a method or getter to compute what counter to show, if any.

Maybe a good structure for that is a method like `Widget? buildTrailing(BuildContext context)`. Then most subclasses will return a `_MenuItemCounter`, or null; but e.g. the "Notifications" row (in the future! it's out of scope for this PR) can return that yellow warning icon.

PIG208 replied:
For the record, we decided to defer the feature to #1088.

gnprice commented on lib/widgets/home.dart:
```diff
@@ -0,0 +1,461 @@
+  }
+}
+
+class NavigationButton extends StatelessWidget {
```
nit: put this just after _HomePageState, since it's closely tied to that

gnprice commented on lib/widgets/home.dart:
```diff
@@ -0,0 +1,461 @@
+    duration: _bottomSheetEnterDuration,
+    reverseDuration: _bottomSheetExitDuration,
+  );
+  bool handleScroll(ScrollNotification notification) {
+    animationController.paused = notification.metrics.extentBefore != 0;
+    return false;
+  }
```
wip; nav: Disable bottom sheet dragging when scrolled

I think the behavior before this commit is fine and actually a bit nicer. So let's leave out this complication.

I gather you're referring to this line in the Figma:
> It would be nice if swipe down can close the menu, but it is important to do that only, when content is scrolled to its top position. And it isn’t just a swipe gesture but full menu dragging 

But I think the thing Vlad means is just: if you've scrolled down in the content and start swiping down, it should scroll you back up in the content (just like normal for a scrollable) instead of having the menu start closing.

When I run with the version before this commit, that all works fine: if you're scrolled down, then swiping anywhere within the scrollable acts on the scrollable before doing anything to the menu as a whole. If you touch the header and drag that, then that does control the menu… but I think that's fine and actually good. After this commit, attempting to drag the header doesn't do anything if the options are scrolled, which to me feels oddly inert like something's not working.

PIG208 replied:
Yeah, I agree to leaving this out.

> It would be nice if swipe down can close the menu

I think to fully implement this (out of scope) feature, we will need to allow over scrolling the inner content (swiping down the menu items and reaching the top of the menu) to drag the bottom sheet; once you scroll pass a certain extent the menu collapses, as it does when you control it by dragging the header. I guess the menu would otherwise rebound back in position if you release it.

It makes sense to leave this out because it will be quite complicated. An implementation might have to listen to notifications like `OverscrollNotification`, `ScrollEndNotification` and replicate the drag handlers from `_BottomSheetState` and manipulate the animation controller.

gnprice reviewed (commented):
Exciting! Looking forward to getting this landed.

Comments below on the draft.

The important high-level point (of which some corollaries appear below) is: the priority with this PR is to get to something we can merge in time for the blog post about the beta.

That means a version that has all the essential functionality, looks reasonable, and meets our usual standards for the code and tests; but any aspects of the design that turn out to be tricky to implement are a lower priority than getting it merged. Because the issue is complex already, anything that's tricky and not essential should be punted until everything essential is complete.

For a concrete timeline: the target for the blog post is by Dec 12, the week after next. I think that calls for a version ready for review by the first half of next week, to give time for review, revisions, and any follow-ups we realize we need after getting to use this a bit.

PIG208 commented on lib/widgets/home.dart:
```diff
@@ -0,0 +1,461 @@
+  }
+}
+
+class _VersionInfo extends StatelessWidget {
```
Opened #1095 as a follow-up.

chrisbobbe commented on lib/widgets/home.dart:
```diff
@@ -2,96 +2,489 @@ import 'package:flutter/material.dart';
+
+    final designVariables = DesignVariables.of(context);
     return Scaffold(
-      appBar: ZulipAppBar(title: const Text("Home")),
```
There's a top bar in the Figma:

<img width="321" alt="image" src="https://github.com/user-attachments/assets/c055a00a-7ba1-40f0-ae39-1dfed37e1a5b">

Ahh but I see #1035 says "There's no top bar" in the issue description. Fair enough 🤷‍♂️ but let's mention it in the commit message's list of differences from the Figma. Also, now we don't have a spec for what goes in the top inset; here's this revision:

| Not scrolled | Scrolled a bit |
| --- | --- |
| <img width="675" alt="image" src="https://github.com/user-attachments/assets/713e959a-2d72-4ed9-859f-66a31c9cfbf5"> | ![image](https://github.com/user-attachments/assets/a4285baa-79ea-4980-9dd4-f5349f37d3b2) |

The top inset is filled with the same color we use under the content in the y-direction when the list is short. I don't love that; it's like the content just abruptly disappears, instead of sliding under some different surface (what used to be the app bar). I don't know what's best here, perhaps just leave a `TODO(design)` for how to fill the top inset on these pages.

PIG208 replied:
Yeah, that will be a separate redesign project. I recalled that we should have an issue for the top bar other than #1039, but I couldn't find it.

gnprice replied:
Let's have an app bar at the top after all. (It'll be pretty plain — just a title, I think.)

I'm not sure what my thinking was when I said in the issue there's no top bar. But I see that's a bit of a challenge what should happen on scroll, if there's no app bar; and I just went and looked through several apps on my phone that have a bottom nav bar, and all of them have an app bar on those screens. In a couple of cases the app bar goes away as you scroll (making more room for the scrolling content), but that's trickier; let's skip that at least for launch, and have the app bar remain fixed at the top on scroll.

gnprice replied:
I went to go update #1035 with this change, and rereading the context of that sentence I realized what happened: what I really meant was that there's no top *tab bar*. I probably didn't consider at that moment the question of whether there should or shouldn't be a plain title-bearing app bar instead.

Anyway, fixed there.

chrisbobbe reviewed (commented):
I've read all the commits except the main commit:

```
nav: Add bottom tabs and main menu
```

which I've skimmed a bit. It's time for bed but I wanted to get a review in today. 🙂

Some things I noticed in manual testing:

- If I tap "Switch account" from the new bottom sheet, then go back (with the back button or iOS swipe gesture), the bottom sheet is still open. Nicer to close it when you tap that button, I think. (Same for some other buttons like Mentions.)

- If I do that after logging out the account I was looking at, I get an infinite loading spinner. This is what `PerAccountStoreWidget.routeToRemoveOnLogout` is meant to prevent. Can we use that directly, or do we need some changes to support auto-dismissing an action sheet on logout?

  <img width="400" alt="image" src="https://github.com/user-attachments/assets/42150226-d592-4bf9-b92b-bb42735dd880">

  If I close that action sheet (by dragging down), what's supposed to happen? I get a black screen:

  <img width="400" alt="image" src="https://github.com/user-attachments/assets/dc70f3fa-7a8e-4559-a2f5-996ea33a1af9">

- When you tap an account on the "Choose account" screen, I notice that a route is added to the stack, on top of the "Choose account" screen, which lets me go back to that screen. Instead, can we clear anything on the stack and replace it with just the bottom-tabs screen for the chosen account? The issue says 'there's no navigating "back" or "up" from [the screen with the bottom tab bar].'

PIG208 commented on assets/l10n/app_en.arb:
```diff
@@ -15,8 +15,8 @@
   },
-  "chooseAccountPageTitle": "Choose account",
-  "@chooseAccountPageTitle": {
+  "switchAccountPageTitle": "Switch account",
```
> Also, what is "i10n", do you mean 'internationalization' ("intl") or localization ("l10n")?

Yeah, that's a typo.

PIG208 commented on assets/l10n/app_en.arb:
```diff
@@ -529,6 +529,10 @@
   },
+  "inboxButton": "Inbox",
+  "@inboxButton": {
+    "description": "Label for the menu button switching to inbox."
```
Updating this with the description altered to match the other existing page titles.

I guess this would also be a reason to also name the button as "Choose account" for the "Choose account" page.

chrisbobbe replied:
> The documentation for `SemanticsProperties.namesRoute` doesn't seem quite clear on how it can be used. I assume that we need to use `Semantics` with it? Because this might be a non-trivial project on its own, I feel that it might fit better as a separate issue.

Yes, definitely fine to leave this for later 🙂:

> How about adding a `TODO(a11y)` to find out if we're meeting users' expectations when they use Zulip with a screen reader, mentioning `SemanticsProperties.namesRoute`?

You could file an issue if you want, but I think the TODO is fine and better than nothing.

PIG208 replied:
I found #535 relevant to this:

> We should try out the app with iOS's VoiceOver and/or Android's TalkBack, and make sure it's possible to do the essentials to be able to use the app at all:

> - get between the different top-level pages (Inbox, All messages, Subscribed streams, Direct messages)

Except that the top-level pages become the tabs on the new home screen. With an app bar it might become less of a concern, but still it would be helpful to experiment with this before launch.

gnprice commented on lib/widgets/page.dart:
```diff
@@ -159,9 +159,6 @@ class LoadingPlaceholderPage extends StatelessWidget {
-      appBar: AppBar(),
-      body: const LoadingPlaceholder(),
-    );
+    return const Scaffold(body: LoadingPlaceholder());
```
The main reason this loading page has an app bar is to make it possible to back out of it (on iOS where you don't have a system back gesture/button) so e.g. you can go use some other account.

With the new nav, backing out wouldn't get you to ChooseAccountPage anyway because HomePage is the root of the navigation. But it'll be important to still have some way to switch accounts from the loading page.

I think that means we want two things here:
* LoadingPlaceholderPage, which is used by AccountPageRouteMixin in general, should continue to have an app bar. That way if you wind up on a page that isn't HomePage, e.g. by opening a notification, and it's stuck loading, you can back out.
* HomePage.buildRoute should provide some alternative loading page that's specific to HomePage. That alternative will provide a way to switch accounts, which can be ad hoc like a button in the bottom right.

PIG208 commented on lib/widgets/home.dart:
```diff
@@ -2,96 +2,489 @@ import 'package:flutter/material.dart';
+          for (final (tab, body) in pageBodies)
+            Offstage(offstage: tab != _tab.value, child: body),
+        ]),
+      bottomNavigationBar: SafeArea(
```
Ah, thanks for testing that out. I found out today that while my Android device has safe area, those area aren't visible at all. Maybe there's still a way to test it without an iOS device?

gnprice replied:
> Maybe there's still a way to test it without an iOS device?

A good way to do that is to use the Android emulator. It has options to simulate a variety of shapes of insets.

chrisbobbe commented:
Test, getting 500s from GitHub when trying to post inline replies above. Ugh.

gnprice commented on lib/widgets/icons.dart:
```diff
@@ -48,65 +48,77 @@ abstract final class ZulipIcons {
   /// The Zulip custom icon "clock".
   static const IconData clock = IconData(0xf108, fontFamily: "Zulip Icons");
 
+  /// The Zulip custom icon "contacts".
+  static const IconData contacts = IconData(0xf109, fontFamily: "Zulip Icons");
```
nit:

    nav: Add icons from Figma
    
    The icons are taken from
      https://www.figma.com/design/1JTNtYo9memgW7vV6d0ygq/Zulip-Mobile?node-id=544-22131&node-type=canvas&t=f4vYPq2CZUpFuf7V-0.

Write instead:

    icons: Add icons for bottom nav bar
    
    The icons are taken from Figma:
      https://www.figma.com/design/1JTNtYo9memgW7vV6d0ygq/Zulip-Mobile?node-id=544-22131&node-type=canvas&t=f4vYPq2CZUpFuf7V-0

Avoiding `.` there avoids ambiguity about what the URL is.

The summary line can be more specific than "nav" about what the icons are; there are other icons involved in nav. And our usual prefix for editing icons is "icons:".

gnprice reviewed (commented):
To help accelerate things for this PR, here's a review of some commits that I think we can merge early. @PIG208 would you send a separate PR with these, after revision for the nit below and applicable comments by @chrisbobbe above?

48671b273 nav [nfc]: Extract *PageBody widgets
70777ba26 nav [nfc]: Move HomePage to a new home
848816262 action_sheet [nfc]: Centralize .withValues variants of contextMenuCancelBg
6d9037b49 nav: Add icons from Figma

In particular I think some of those are likely to be conflict-prone — icons changes always are, and theme changes often seem to be — so it's helpful to have them merged early.

gnprice commented on test/widgets/home_test.dart:
```diff
@@ -0,0 +1,155 @@
+      await tapOpenMenu(tester);
+      checkIconSelected(tester, inboxMenuIconFinder);
+      checkIconNotSelected(tester, channelsMenuIconFinder);
+      check(find.byType(InboxPageBody)).findsOne();
```
```suggestion
      check(find.byType(InboxPageBody)).findsOne();
      check(find.byType(SubscriptionListPageBody)).findsNothing();
```

and similarly `check(find.byType(InboxPageBody)).findsNothing();` below. That helps the reader be sure that something really changed as a result of the tap — that it's not just that both finders already would find something and continue to do so (e.g. because we got something wrong in the interaction of the finders with the `Offstage` nature of the pages).

gnprice reviewed (commented):
Further in the interest of accelerating things for this PR, here's an initial review from me of the main commit:
6e8de9178 nav: Add bottom tabs and main menu

I skipped points that are already discussed in Chris's review above.

PIG208 commented on lib/widgets/home.dart:
```diff
@@ -0,0 +1,490 @@
+          child: Align(
+            // This is necessary to limit the width of the row of icons,
+            // because [Align] allows the child to be smaller than itself.
+            heightFactor: 1,
```
Moved this comment above and added another one explaining the `heightFactor`. Also changed it to use `Center` instead.

PIG208 commented on lib/widgets/home.dart:
```diff
@@ -0,0 +1,490 @@
+                  duration: const Duration(milliseconds: 100),
+                  child: ConstrainedBox(
+                    constraints: const BoxConstraints(minHeight: 44),
+                    child: const ActionSheetCancelButton()))),
```
With a smaller font, the button can get a bit shorter than that:

<details>
<summary>Small font</summary>

| w/ minHeight | w/o minHeight |
| - | - |
| ![1000015714](https://github.com/user-attachments/assets/d66dbc5d-f580-4ab3-9496-e1105e5acf49) | ![1000015713](https://github.com/user-attachments/assets/ca863c21-e62e-4012-b850-68ec1aeebec2) |


</details>

gnprice replied:
(Are the labels switched on those screenshots? The one labeled "w/ minHeight" has the button shorter than the other does.)

gnprice replied:
Following up on this point:

> I'd think TextButton (which this button uses) would take care of that.

by looking into TextButton, it looks like this is controlled by `ButtonStyle.minimumSize`. The default for which in `TextButton.defaultStyleOf` is:
```dart
  @override
  MaterialStateProperty<Size>? get minimumSize =>
    const MaterialStatePropertyAll<Size>(Size(64.0, 40.0));
```

That's a minimum of 64px wide but only 40px high. So TextButton does indeed take care of this need — but it sets a slightly smaller minimum.

What if we set `TextButton.style` to provide a `minimumSize` property there?

PIG208 replied:
> What if we set TextButton.style to provide a minimumSize property there?

Yeah, that should be cleaner.

<details>
<summary>Updated screenshots</summary>


| font size | before | after |
| - | - | - |
| extra small | ![small_without_fix](https://github.com/user-attachments/assets/cb45c45b-28c4-46dd-9312-1d742e02aae8) | ![small_with_fix](https://github.com/user-attachments/assets/4219cfb6-3348-4d28-a0e9-42aa25fd6085)
| normal |  ![normal](https://github.com/user-attachments/assets/7c871cad-3e20-4a1f-85ea-b31c1fdc484e) | / |

</details>

PIG208 commented:
Started experimenting with the special placeholder and route changes.

This addresses most of the comment except
https://github.com/zulip/zulip-flutter/pull/1076#discussion_r1868262551 (special loading page) and https://github.com/zulip/zulip-flutter/pull/1076#pullrequestreview-2474679726 (navigation change) and https://github.com/zulip/zulip-flutter/pull/1076#discussion_r1868288529 (app bar). All of these mentioned features are included as WIP for now. Will get back to this tomorrow.

chrisbobbe commented:
The prep PR #1099 was merged, so this can be updated for that. (Also I see CI is failing.)

PIG208 commented:
Yes. There are some wip changes to be finished before this is ready for review again.

PIG208 commented:
Thanks for the reviews @chrisbobbe @gnprice! The PR is ready again.

Some UI changes:

<details>
<summary>HomePage's loading screen</summary>

![Screenshot_20241205_010236](https://github.com/user-attachments/assets/40b62d6c-c2a8-456b-9978-1b2c41927f18)

-
![Screenshot_20241205_010241](https://github.com/user-attachments/assets/1f97b655-d6fe-4b65-8354-fcede6352088)


</details>

<details>
<summary>HomePage's app bar</summary>

![Screenshot_20241205_005926(1)](https://github.com/user-attachments/assets/8f75f0fa-0e99-426f-a9b4-14940d217342)

</details>

Now we ensure that the HomePage is always at the root level, by removing all the existing routes when pushing it onto the navigator stack.

There are also some minor changes like popping the bottom sheet for all menu buttons.

gnprice commented on lib/widgets/home.dart:
```diff
@@ -2,96 +2,533 @@ import 'package:flutter/material.dart';
+  }
+}
+
+class _NavigationButton extends StatelessWidget {
```
bump https://github.com/zulip/zulip-flutter/pull/1076#discussion_r1868535479

PIG208 replied:
I see. Missed the `Bar` part. Fixed that and updated comment/string references to this name to match.

gnprice reviewed (commented):
Thanks for the revision!

It's late, so for tonight here's just a review of the points that were in my previous review https://github.com/zulip/zulip-flutter/pull/1076#pullrequestreview-2477006695 . All look good now except for the couple of items below.

gnprice commented on test/widgets/login_test.dart:
```diff
@@ -118,30 +123,71 @@ void main() {
+        pushedRoutes.clear();
+
+        await login(tester, eg.otherAccount);
+        final loggedInAccount = testBinding.globalStore.accounts.singleWhere(
```
nit: "logged-in account" is a confusing name here, because both accounts are logged in.

```suggestion
        final newAccount = testBinding.globalStore.accounts.singleWhere(
```

gnprice commented on lib/widgets/home.dart:
```diff
@@ -2,96 +2,532 @@ import 'package:flutter/material.dart';
-                    narrow: ChannelNarrow(testStreamId!))),
-                child: const Text("#test here")), // scaffolding hack, see above
-            ],
+      body: Column(
+        crossAxisAlignment: CrossAxisAlignment.end,
+        children: [
+          const Expanded(child: LoadingPlaceholder()),
+          TextButton(
```
This needs to use SafeArea to stay out of the insets. Here's how it looks on my phone:
![image](https://github.com/user-attachments/assets/aa4231f6-2533-4d87-b10e-79faae13826d)

gnprice replied:
Probably cleanest is to make this a floating action button:
```dart
    return Scaffold(
      body: const LoadingPlaceholder(),
      floatingActionButton: TextButton(
        onPressed: () => Navigator.push(context,
          MaterialWidgetRoute(page: const ChooseAccountPage())),
        child: Text(zulipLocalizations.chooseAccountPageTitle)),
    );
```

That takes care of applying SafeArea; and it also means the layout of LoadingPlaceholder and centering of the circular progress indicator doesn't get distorted by the presence of the button.

PIG208 replied:
In this revision I also added the `AppBar` back because `HomePage` has gained one in the previous update.

gnprice reviewed (commented):
Thanks for the revision! I played with this UI and am happy to see that the behavior seems to be basically all working.

Partial review below, including one behavior issue in the switch-account button being potentially inaccessible. Posting the partial review now in the interest of pipelining work on this PR.

Another thing I noticed in the UI is that the menu takes up the full height of the screen even when its contents are much shorter. Is that something Vlad specifically addressed in his design? My instinct would be to let the menu be the height of its contents when there's room, and fill the screen only when it needs to.

PIG208 commented on test/widgets/app_test.dart:
```diff
@@ -23,39 +24,80 @@ void main() {
+      ]);
+    });
+
+    testWidgets('choosing an account clears the navigator stack', (tester) async {
```
Moving this to the "ChooseAccountPage" group.

PIG208 commented:
> Another thing I noticed in the UI is that the menu takes up the full height of the screen even when its contents are much shorter. Is that something Vlad specifically addressed in his design? My instinct would be to let the menu be the height of its contents when there's room, and fill the screen only when it needs to.

Yeah, it's mentioned [here](https://www.figma.com/design/1JTNtYo9memgW7vV6d0ygq/Zulip-Mobile?node-id=2957-52907&node-type=frame&t=LagtQe63trPehowh-0):

> Menu slides up to full screen state. There is not medium state, and no dragging icon. 

I interpret this as "the menu always fills the entire screen".

chrisbobbe commented on lib/widgets/home.dart:
```diff
@@ -2,96 +2,528 @@ import 'package:flutter/material.dart';
+      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(10)),
+    ).copyWith(
+      backgroundColor: WidgetStateColor.fromMap({
+        WidgetState.pressed: designVariables.bgMenuButtonActive,
```
In dark mode, the pressed button background `bgMenuButtonActive` is black with 20% opacity, and it's on a `#222222` background (`bgBotBar`). So it should be darker than the background. But it's not:

<img width="664" alt="image" src="https://github.com/user-attachments/assets/f24bb043-9a2e-420b-b52b-b12212635e1b">

…well actually it is 🙂 but only for a frame or two before an animation makes it light like this. I think the culprit is a default `overlayColor` value or something like that. I notice that it gets darker if I pass `overlayColor: Colors.transparent`; that may or may not be the solution but it's an observation.

PIG208 replied:
I found it helpful to look at the defaults:

```
  /// ## Material 3 defaults
  ///
  /// If [ThemeData.useMaterial3] is set to true the following defaults will
  /// be used:
  ///
  /// {@template flutter.material.text_button.material3_defaults}
  /// * `textStyle` - Theme.textTheme.labelLarge
  /// * `backgroundColor` - transparent
  /// * `foregroundColor`
  ///   * disabled - Theme.colorScheme.onSurface(0.38)
  ///   * others - Theme.colorScheme.primary
  /// * `overlayColor`
  ///   * hovered - Theme.colorScheme.primary(0.08)
  ///   * focused or pressed - Theme.colorScheme.primary(0.1)
  ///   * others - null
  /// * `shadowColor` - Colors.transparent,
```
I guess visually the `pressed` state got affected because `overlayColor` is `Theme.colorScheme.primary(0.1)` by default. We will lose the color distinction for the `hovered`/`focused` state, but we can still control them if we need to.

chrisbobbe reviewed (changes requested):
Thanks! Here's another not-quite-thorough review before my bedtime. I still haven't read the tests in the main commit.

I notice you have a commit marked "fixup!" so it's possible this wasn't ready for another review?

I notice this isn't ready to support the situation where the app logs you out without you requesting it in the UI:

https://github.com/zulip/zulip-flutter/pull/1076#pullrequestreview-2474679726
> This way might be nice because it also supports cases where you logged out without requesting it in the UI, like (not yet implemented) when we learn your auth is invalid.

I tested by just changing the "Inbox" menu's press handler so that it called `logOutAccount`, and when I pressed it, it put me on a blank black screen. Since that case isn't currently reachable (right now the only way to be logged out is to request it from the choose-account page), it might be fine to leave for a followup.

I remember one of your commits said this is just a partial implementation, with more things planned as followups—do we have a list of those things somewhere?

gnprice commented on lib/widgets/home.dart:
```diff
@@ -2,96 +2,528 @@ import 'package:flutter/material.dart';
-                  MessageListPage.buildRoute(context: context,
-                    narrow: ChannelNarrow(testStreamId!))),
-                child: const Text("#test here")), // scaffolding hack, see above
-            ],
+      appBar: AppBar(),
+      body: const LoadingPlaceholder(),
+      floatingActionButton: TextButton(
+        onPressed: () => Navigator.push(context,
+          MaterialWidgetRoute(page: const ChooseAccountPage())),
+        child: Text(zulipLocalizations.chooseAccountPageTitle)));
+  }
```
Yeah, "Choose account" isn't good text here. You made the same point about the button in the menu and I agree. Better text would be "Switch account", or perhaps like "Try another account".

The reason something like this is needed is in case the data for this account takes a long time to load — perhaps isn't ever going to load, because the server is down — so the user can still use some other account if that's what they wanted.

PIG208 replied:
In addition to using "Try another account", does it look better as a `FilledButton` or an `OutlinedButton`? (so that user knows that it is a button)

| `TextButton` | `OutlinedButton` | `FilledButton` |
| - | - | - |
| ![Screenshot from 2024-12-06 15-09-11](https://github.com/user-attachments/assets/ff882c7d-1c5f-4de8-a924-17154e5164ae) | ![Screenshot from 2024-12-06 15-09-03](https://github.com/user-attachments/assets/88073307-90d5-41e9-8a92-c5dc1b60e835) | ![Screenshot from 2024-12-06 15-08-51](https://github.com/user-attachments/assets/202d855e-78e4-4dac-ae86-9d3d5cae464f) |

chrisbobbe replied:
Yeah, I think the button's current appearance doesn't really read as a button.

Maybe a nicer alternative (might take longer to implement) would be to wait 5 seconds or so before showing anything, and have it be a message (centered in the page?) like "This is taking longer than usual to load. Try another account?" That way in the happy-path case we don't distract the user by making them figure out UI that they won't need.

PIG208 replied:
Started a discussion [here](https://chat.zulip.org/#narrow/channel/48-mobile/topic/.23F1076.20accessing.20choose-account.20page.20from.20the.20loading.20screen).

PIG208 commented on lib/widgets/app.dart:
```diff
@@ -209,11 +208,10 @@ class _ZulipAppState extends State<ZulipApp> with WidgetsBindingObserver {
-              MaterialWidgetRoute(page: const ChooseAccountPage()),
-              if (initialAccountId != null) ...[
+              if (initialAccountId != null)
                 HomePage.buildRoute(accountId: initialAccountId),
-                InboxPage.buildRoute(accountId: initialAccountId),
-              ],
+              if (initialAccountId == null)
+                MaterialWidgetRoute(page: const ChooseAccountPage()),
             ];
```
Ah, I tried that before and was wondering why it wasn't working. It turned out that I had the comma at the end of the if expression.

gnprice commented on lib/widgets/home.dart:
```diff
@@ -2,96 +2,528 @@ import 'package:flutter/material.dart';
+    clipBehavior: Clip.antiAlias,
+    useSafeArea: true,
+    isScrollControlled: true,
+    backgroundColor: designVariables.bgBotBar,
```
Hmm yeah, that's a bug.

I don't see a good way to avoid it with the existing API of `showModalBottomSheet` and friends, though. (If we wanted the same custom background color for all our bottom sheets, or all our modal bottom sheets, we could do that with ThemeData.bottomSheetTheme. But this has a different background color from the message/topic/etc. action sheets.) So let's file that as a follow-up issue for post-launch.

PIG208 replied:
If we do not pass `backgroundColor` on the `showModalBottomSheet` call, we can set the background color on a `DecoratedBox` and have the `DesignVariables.of(context)`'s inline, or do something like:

```dart
  showModalBottomSheet<void>(
    // [...]
    builder: (BuildContext context) {
      final designVariables = DesignVariables.of(context);
      return PerAccountStoreWidget(
        accountId: accountId,
        child: DecoratedBox(
          decoration: BoxDecoration(color: designVariables.bgBotBar),
```

This does not set the color of the bottom sheet `Material`. So we also have to set `badckgroundColor` to something like `transparent` to avoid rendering artifacts:

![image](https://github.com/user-attachments/assets/992740eb-0555-4cbe-8d7a-4be62c90ce3b)

(with `backgroundColor: Colors.blue`)

gnprice replied:
Yeah, and the `Material` also gets involved (and uses its color) in a variety of interactions, which we'd effectively be turning off. So that feels like a workaround approach that's likely to have subtle bugs of its own. I'm OK with just having that glitch if the theme happens to change just at the moment that you have the main menu open, and leaving the fix for it as a future issue.

gnprice commented on lib/widgets/page.dart:
```diff
@@ -152,6 +159,9 @@ class AccountPageRouteBuilder<T extends Object?> extends PageRouteBuilder<T> wit
   final int accountId;
+
+  @override
+  Widget? loadingPlaceholderPage;
```
```suggestion
  final Widget? loadingPlaceholderPage;
```

gnprice commented on test/widgets/actions_test.dart:
```diff
@@ -182,8 +182,10 @@ void main() {
       testNavObserver.onPushed = (route, prevRoute) => pushedRoutes.add(route);
-      final account1Route = InboxPage.buildRoute(accountId: account1.id);
-      final account2Route = InboxPage.buildRoute(accountId: account2.id);
+      final account1Route = MaterialAccountWidgetRoute(
+        accountId: account1.id, page: const InboxPageBody());
+      final account2Route = MaterialAccountWidgetRoute(
+        accountId: account2.id, page: const InboxPageBody());
```
These pages don't exist in real life. That rather undermines this test, because it's all about how the navigation stack will behave when one account gets logged out.

… In fact I think contemplating what a version of this test would look like that's updated to be realistic after this PR leads to rediscovering the bug that @chrisbobbe mentioned at https://github.com/zulip/zulip-flutter/pull/1076#pullrequestreview-2474679726 and https://github.com/zulip/zulip-flutter/pull/1076#pullrequestreview-2483685141 .

As Chris said in that latter comment (and I agree), that issue can be left as a followup, because it's not a live bug until we make other changes in the future. So I guess this test can just get a TODO comment pointing at that followup issue.

gnprice replied:
bump: should get that TODO comment

gnprice commented on test/widgets/recent_dm_conversations_test.dart:
```diff
@@ -49,7 +50,7 @@ Future<void> setupPage(WidgetTester tester, {
   await tester.pumpWidget(TestZulipApp(
     accountId: eg.selfAccount.id,
     navigatorObservers: navigatorObserver != null ? [navigatorObserver] : [],
-    child: const RecentDmConversationsPage()));
+    child: const HomePage(initialTab: HomePageTab.directMessages)));
```
How about making this just `HomePage()`, and following with a `tester.tap(find.byIcon(…))` to get to the right tab?

That'd be more realistic, and (if we do the same for the subscription/channels tab below) would let us avoid having this `initialTab` field that's used only in tests.

gnprice commented on lib/widgets/app.dart:
```diff
@@ -209,11 +208,10 @@ class _ZulipAppState extends State<ZulipApp> with WidgetsBindingObserver {
                 HomePage.buildRoute(accountId: initialAccountId),
-                InboxPage.buildRoute(accountId: initialAccountId),
-              ],
+              if (initialAccountId == null)
+                MaterialWidgetRoute(page: const ChooseAccountPage()),
```
nit: put the no-account case first; it's the case that always happens first chronologically, and it's also the unusual case so going first is consistent with our usual early-return structure

gnprice commented on assets/l10n/app_en.arb:
```diff
@@ -529,6 +533,10 @@
   "@userRoleUnknown": {
     "description": "Label for UserRole.unknown"
   },
+  "inboxPageTitle": "Inbox",
+  "@inboxPageTitle": {
+    "description": "Title for the page of inbox."
```
This description doesn't read very well to me. But then I realized the same is true of some of the existing descriptions of similar strings, so I sent a quick PR fixing those:
* #1109 

This can then follow the style of those.

gnprice commented on assets/l10n/app_en.arb:
```diff
@@ -19,6 +19,10 @@
   },
+  "switchAccountButton": "Switch account",
+  "@switchAccountButton": {
+    "description": "Main menu button for the choose-account page"
```
```suggestion
    "description": "Label for main-menu button leading to the choose-account page."
```

The string isn't a button, but rather the label for a button.

Separately, without the hyphen it reads to me a bit more like: it's (a label for) a menu button for the choose-account page, namely the main such button.

gnprice commented on assets/l10n/app_en.arb:
```diff
@@ -545,6 +553,14 @@
+  "@channelsPageTitle": {
+    "description": "Title for the page of channels."
+  },
+  "profilePageTitle": "My profile",
+  "@profilePageTitle": {
+    "description": "Title for the page of the logged in user's profile."
```
This name and description is inaccurate — the PR doesn't use this string for the title of the page. The page's title is what appears in the app bar, which for this page is the user's name.

Instead a good description would be:
> Label for main-menu button leading to the user's own profile.

And for the name… perhaps `mainMenuMyProfile`.

gnprice commented on lib/widgets/app.dart:
```diff
@@ -268,17 +266,18 @@ class ChooseAccountPage extends StatelessWidget {
     final zulipLocalizations = ZulipLocalizations.of(context);
     assert(!PerAccountStoreWidget.debugExistsOf(context));
     final globalStore = GlobalStoreWidget.of(context);
+    final hasBackButton = ModalRoute.of(context)?.impliesAppBarDismissal ?? false;
```
This line is pretty mysterious — let's give it a one- or two-line comment pointing at how you determined this was the right way to compute whether there'll be a back button.

gnprice commented on lib/widgets/home.dart:
```diff
@@ -1,97 +1,539 @@
+class _HomePageState extends State<HomePage> {
+  late final ValueNotifier<HomePageTab> _tab = ValueNotifier(widget.initialTab);
+
+  String get _currentTabTitle {
```
nit: put this just above `build`, after the state management; it's effectively a private helper for `build`

gnprice commented on lib/widgets/home.dart:
```diff
@@ -1,97 +1,539 @@
+          border: Border(top: BorderSide(color: designVariables.borderBar, width: 0)),
+          color: designVariables.bgBotBar),
+        child: SafeArea(
+          // [Center] is necessary to limit the width of the row of icons,
+          // because it allows the child to be smaller than itself.
+          child: Center(
+            // This shrinks [Center] to be as tall as its child, instead of
+            // growing as tall as [Scaffold] allows, filling the entire screen.
+            heightFactor: 1,
+            child: ConstrainedBox(
+              // TODO(design): determine a suitable max width for bottom nav bar
+              constraints: const BoxConstraints(maxWidth: 600).tighten(height: 48),
```
It's kept bugging me that this feels complicated, so I fiddled a bit to find a simpler version:
```suggestion
          child: SizedBox(height: 48,
            child: Center(
              child: ConstrainedBox(
                // TODO(design): determine a suitable max width for bottom nav bar
                constraints: const BoxConstraints(maxWidth: 600),
```

(The `CrossAxisAlignment.stretch` means the child will indeed fill those 48px of height, and therefore so will the Center widget)

gnprice commented on lib/widgets/home.dart:
```diff
@@ -1,97 +1,539 @@
+  }
+}
+
+void _showMenu(BuildContext context, {
```
nit:
```suggestion
void _showMainMenu(BuildContext context, {
```

Makes it clearer up front which menu it is that this function shows.

gnprice commented on lib/widgets/home.dart:
```diff
@@ -1,97 +1,539 @@
+  const _SwitchAccountButton();
+
+  @override
+  // TODO(design): choose an icon
+  IconData get icon => ZulipIcons.globe;
```
Yeah, not `globe` because that's the icon we use to mean "web-public channel".

The TODO can remain, but pending that it'd be better to just make this blank — have `buildLeading` return `SizedBox.shrink()`.

PIG208 commented:
(will update tests for the loading screen shortly)

PIG208 commented:
<details>
<summary>Light mode</summary>

| before | after |
| - | - |
| ![light-1](https://github.com/user-attachments/assets/eb5bb70c-a4e0-4854-829c-af430e0f0295) | ![light-1](https://github.com/user-attachments/assets/548b14c2-7af2-4bcc-ba90-684c39c2a053) |
| ![light-2](https://github.com/user-attachments/assets/b2481948-56b8-4d75-ba2f-3fb0853affc1) | ![light-2](https://github.com/user-attachments/assets/b045a16f-74e2-4ec1-b952-6a4bc865dc12) |
| ![light-3](https://github.com/user-attachments/assets/9a5cab4f-0aa4-4564-b88d-c0d13158fa2a) | ![light-3](https://github.com/user-attachments/assets/56b41dac-de10-40f4-b39b-7242c74940cc) |
| ![light-4](https://github.com/user-attachments/assets/98447dd0-13ce-4ee3-a098-6a8bdddbc3e1) | ![light-4](https://github.com/user-attachments/assets/e681a523-68ff-41e6-a221-d30f975e09ed) |
| ![light-5](https://github.com/user-attachments/assets/86ee3ad6-1dbc-4925-82e1-00e3b2237aa7) | ![light-5](https://github.com/user-attachments/assets/0f038706-ffbe-4b45-aee2-9aa1a7bc6e0c) |
| ![light-6](https://github.com/user-attachments/assets/3578e816-0d1f-480a-9699-17406f6dba84) | ![light-6](https://github.com/user-attachments/assets/e41e3d10-97f8-43ba-9a6e-4747061922aa) |

</details>

<details>
<summary>Dark mode</summary>

| before | after |
| - | - |
|  ![dark-1](https://github.com/user-attachments/assets/4f502412-aa62-4b44-a042-1f756c7b8e1a) | ![dark-1](https://github.com/user-attachments/assets/df4cb3e4-9e97-4662-9a8f-2eff07b8d3b2) |
| ![dark-2](https://github.com/user-attachments/assets/9a7e921a-d10c-4590-8754-a76be8b805e4) | ![dark-2](https://github.com/user-attachments/assets/783d2ffa-8e31-410f-8e41-0151b23a78a3) |
| ![dark-4](https://github.com/user-attachments/assets/2bb1205d-b095-4fc1-b9f0-ebc5d890fa8d) | ![dark-3](https://github.com/user-attachments/assets/c85210a6-0c0f-4ba9-b5f0-94f4aef93773) |
| ![dark-3](https://github.com/user-attachments/assets/75c6e3e2-7879-49f2-a414-e69e9a9a9288) | ![dark-4](https://github.com/user-attachments/assets/bfd79753-d341-4aec-aeda-945e8a3ad562) |
| ![dark-5](https://github.com/user-attachments/assets/0bf170b2-d763-45aa-870a-808a07a66175) | ![dark-5](https://github.com/user-attachments/assets/52284d73-ee3a-48e8-8571-5628cc366edb) |
| ![dark-6](https://github.com/user-attachments/assets/6ac1390e-b5c5-4d7e-9ec2-dcc88c3aad33) | ![dark-6](https://github.com/user-attachments/assets/2c664bb5-aa4d-421f-b200-9c2b8b5b5e6c) |


</details>

PIG208 commented:
Looks like `SafeArea` was dropped (`git diff b92737fe..0037df92 -- lib/widgets/home.dart`). Adding it back.

gnprice commented on lib/widgets/theme.dart:
```diff
@@ -82,10 +89,7 @@ ThemeData zulipThemeData(BuildContext context) {
     // Or try this tool to see the whole palette:
     //   https://m3.material.io/theme-builder#/custom
-    colorScheme: ColorScheme.fromSeed(
-      brightness: brightness,
-      seedColor: kZulipBrandColor,
-    ),
+    colorScheme: colorScheme,
```
nit: the comment above here is about the construction of the ColorScheme, so should move along with that constructor call

gnprice reviewed (commented):
Thanks for the revision!

I did a quick re-review of mostly just the items that were in my last review; comments from that below, and otherwise those items all look good.

(These are minor and this is definitely lower-priority than the investigation you're doing at the moment on #924 — posting it now just to flush the state from my head and my GitHub pending comments.)

gnprice commented on assets/l10n/app_en.arb:
```diff
@@ -17,7 +17,20 @@
+  "@switchAccountButton": {
+    "description": "Label for main-menu button leading to the choose-account page."
+  },
+  "tryAnotherAccountMessage": "You account at {url} is taking a while to load.",
```
```suggestion
  "tryAnotherAccountMessage": "Your account at {url} is taking a while to load.",
```

gnprice commented on assets/l10n/app_en.arb:
```diff
@@ -17,7 +17,20 @@
+  "@switchAccountButton": {
+    "description": "Label for main-menu button leading to the choose-account page."
+  },
+  "tryAnotherAccountMessage": "You account at {url} is taking a while to load.",
+  "@tryAnotherAccountMessage": {
+    "description": "Message that appears on the loading screen after waiting for some time.",
+    "url": {"type": "String", "example": "http://example.com/"}
```
nit:
```suggestion
  "tryAnotherAccountMessage": "You account at {url} is taking a while to load.",
  "@tryAnotherAccountMessage": {
    "description": "Message that appears on the loading screen after waiting for some time.",
    "url": {"type": "String", "example": "http://chat.example.com/"}
```

gnprice commented on lib/widgets/app.dart:
```diff
@@ -268,41 +266,50 @@ class ChooseAccountPage extends StatelessWidget {
 
   @override
   Widget build(BuildContext context) {
+    final colorScheme = ColorScheme.of(context);
```
This commit also edits colors for the individual `ListTile`s above:
```diff
       child: ListTile(
         title: title,
         subtitle: subtitle,
+        tileColor: colorScheme.secondaryContainer,
+        textColor: colorScheme.onSecondaryContainer,
```

but then the main commit reverts that.

It doesn't look like the main commit is making some other change that compensates for removing it, so that change probably doesn't belong in the main commit. Are those lines not actually needed in the first place?

PIG208 replied:
Those lines should be kept to fix the appearance of the account cards on choose-acocunt page.

gnprice commented on lib/widgets/app.dart:
```diff
@@ -268,41 +266,50 @@ class ChooseAccountPage extends StatelessWidget {
-                child: Text(zulipLocalizations.chooseAccountButtonAddAnAccount)),
-            ]))),
-      ));
+    // Borrowed from [AppBar.build].
+    // See documentation on [ModalRoute.impliesAppBarDismissal]:
+    /// > Whether an [AppBar] in the route should automatically add a back button or
+    /// > close button.
+    final hasBackButton = ModalRoute.of(context)?.impliesAppBarDismissal ?? false;
```
nit: set off this stanza with blank lines:
```suggestion

    // Borrowed from [AppBar.build].
    // See documentation on [ModalRoute.impliesAppBarDismissal]:
    /// > Whether an [AppBar] in the route should automatically add a back button or
    /// > close button.
    final hasBackButton = ModalRoute.of(context)?.impliesAppBarDismissal ?? false;

```

In particular that helps make clear what code the comment is describing.

gnprice replied:
nit: those `///` should be `//`, right?

gnprice reviewed (commented):
And here's a full review… up to partway through the main commit, just before starting on `home.dart`. (So not covering the bulk of the PR.) Posting now for pipelining, as today's a day where I'm switching between multiple tasks.

alya commented:
The "Log in with X" buttons look really squashed together in the screenshots above. Should we add a bit of vertical space?

PIG208 commented:
Looks like we are not getting the padded buttons on the Linux build because the buttons are padded to 48px in size by default on mobile devices:


<details>
<summary>Details</summary>

| light | dark |
| - | - |
| ![1000015830](https://github.com/user-attachments/assets/967e1d2d-7ad1-444b-ba32-0898d1e70597) | ![1000015828](https://github.com/user-attachments/assets/bbcfe36e-9e92-41ed-9500-3bb2ef8c0c49) |

</details>

We can override that default so it is padded explicitly on all platforms.

gnprice commented on test/widgets/app_test.dart:
```diff
@@ -162,6 +159,45 @@ void main() {
+      await testBinding.globalStore.add(eg.selfAccount, eg.initialSnapshot());
+      await testBinding.globalStore.add(eg.otherAccount, eg.initialSnapshot());
+
+      final pushedRoutes = <Route<dynamic>>[];
+      final poppedRoutes = <Route<dynamic>>[];
```
nit: avoid `dynamic`:
```suggestion
      final pushedRoutes = <Route<void>>[];
      final poppedRoutes = <Route<void>>[];
```

gnprice commented on test/widgets/login_test.dart:
```diff
@@ -118,30 +124,57 @@ void main() {
             id: testBinding.globalStore.accounts.single.id));
       });
 
+      testWidgets('logging into a different account', (tester) async {
```
nit:
```suggestion
      testWidgets('logging into a second account', (tester) async {
```

Otherwise it raises the question, different from what? The point of this particular test is that it's one where there was already an account when logging into another one.

gnprice commented on test/widgets/home_test.dart:
```diff
@@ -0,0 +1,409 @@
+void main () {
+  TestZulipBinding.ensureInitialized();
+
+  late PerAccountStore store;
```
nit: make this final non-late, inside `prepare` — we don't actually use having it shared like this

PIG208 replied:
Skipping this for https://github.com/zulip/zulip-flutter/pull/1076#discussion_r1879243089.

gnprice commented on test/widgets/home_test.dart:
```diff
@@ -0,0 +1,409 @@
+    await store.addStream(stream);
+    await store.addSubscription(eg.subscription(stream));
+
+    await store.handleEvent(MessageEvent(
+      id: 0, message: eg.streamMessage(stream: stream, sender: eg.otherUser)));
```
This feels awfully specific. Is it generally needed?

Looks like just one test fails if I delete it. Can that test do this for itself?

PIG208 replied:
It's for preparing a message in the inbox. We can either access `PerAccountStore` from that test, keep the `store.handleEvent` inside, or find a different state to test.

gnprice replied:
Yeah, accessing the store from that test sounds good.

gnprice replied:
(and then I guess it'll stay a late, shared variable)

gnprice commented on test/widgets/home_test.dart:
```diff
@@ -0,0 +1,409 @@
+
+    await tester.pumpWidget(TestZulipApp(
+      accountId: eg.selfAccount.id, child: const HomePage()));
+    await tester.pumpAndSettle();
```
Avoid pumpAndSettle:
```suggestion
    await tester.pump();
```

https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-using-pumpandsettle

gnprice commented on test/widgets/home_test.dart:
```diff
@@ -0,0 +1,409 @@
+  group('menu', () {
+    final designVariables = DesignVariables.light();
+
+    final inboxMenuIconFinder = find.descendant(
+      of: find.byType(SingleChildScrollView),
+      matching: find.byIcon(ZulipIcons.inbox));
```
How about:
```suggestion
    final inboxMenuIconFinder = find.descendant(
      of: find.byType(BottomSheet),
      matching: find.byIcon(ZulipIcons.inbox));
```

That feels a bit closer to the UI as seen by the user. After all, for users not on small screens the menu doesn't even scroll.

(The point of the ancestor condition IIUC is to exclude the corresponding icon in the nav bar itself.)

gnprice commented on test/widgets/home_test.dart:
```diff
@@ -0,0 +1,409 @@
+      check(find.byType(HomePage)).findsNothing();
+    }
+
+    ModalRoute<dynamic>? getRouteOf(WidgetTester tester, Finder elementFinder) =>
+      ModalRoute.of(tester.element(elementFinder));
```
nit: it's just the normal type of finder, no need to be more specific than usual:
```suggestion
    ModalRoute<void>? getRouteOf(WidgetTester tester, Finder finder) =>
      ModalRoute.of(tester.element(finder));
```

(and separately, avoid dynamic)

gnprice commented on test/widgets/home_test.dart:
```diff
@@ -0,0 +1,409 @@
+      checkOnHomePage(tester, expectedAccount: eg.selfAccount);
+    });
+
+    testWidgets('while loading, go back from ChooseAccountPage', (tester) async {
+      testBinding.globalStore.loadPerAccountDuration = loadPerAccountDuration;
+      await prepare(tester);
+      await tester.pump(kTryAnotherAccountWaitPeriod);
+      await tapChooseAccount(tester);
+
+      await tester.tap(find.byType(BackButton));
+      await tester.pump(Duration.zero); // tap the button
+      await tester.pump(const Duration(milliseconds: 350)); // wait for pop animation
+      checkOnLoadingPage();
+
+      await tester.pump(loadPerAccountDuration);
+      checkOnHomePage(tester, expectedAccount: eg.selfAccount);
+    });
```
I like these nice clear tests! The various helper functions really contribute to being able to tell these different stories cleanly.

gnprice commented on test/widgets/home_test.dart:
```diff
@@ -0,0 +1,409 @@
+      await tester.pump(loadPerAccountDuration);
+      checkOnLoadingPage();
+
+      // The second loadPerAccount finished.
+      await tester.pump(loadPerAccountDuration);
+      checkOnHomePage(tester, expectedAccount: eg.otherAccount);
```
nit: comment that says something is true, in the present tense, should go at a point where it's become true

```suggestion
      await tester.pump(loadPerAccountDuration);
      // The second loadPerAccount finished.
      checkOnHomePage(tester, expectedAccount: eg.otherAccount);
```

gnprice commented on test/widgets/home_test.dart:
```diff
@@ -0,0 +1,409 @@
+      testBinding.globalStore.loadPerAccountDuration = loadPerAccountDuration * 2;
+      await chooseAccountWithEmail(tester, eg.otherAccount.email);
+
+      // The second loadPerAccount is still pending.
+      await tester.pump(loadPerAccountDuration);
+      checkOnLoadingPage();
```
nit: similarly:
```suggestion
      await tester.pump(loadPerAccountDuration);
      // The second loadPerAccount is still pending.
      checkOnLoadingPage();
```

Here the statement in the comment was already true before the pump, but it wasn't interesting — of course it's still pending, only a few seconds have passed. It's after the await that that fact becomes interesting (which corresponds to the fact that that's the point where you're checking it).

gnprice commented on test/widgets/home_test.dart:
```diff
@@ -0,0 +1,409 @@
+      checkOnHomePage(tester, expectedAccount: eg.otherAccount);
+    });
+
+    testWidgets('while loading, choosing an account disallow going back', (tester) async {
```
nit:
```suggestion
    testWidgets('while loading, choosing an account disallows going back', (tester) async {
```

gnprice commented on test/widgets/recent_dm_conversations_test.dart:
```diff
@@ -49,10 +50,16 @@ Future<void> setupPage(WidgetTester tester, {
   // global store, per-account store, and page get loaded
   await tester.pumpAndSettle();
+
+  // Switch to direct messages tab.
+  await tester.tap(find.descendant(
+    of: find.byType(Center),
+    matching: find.byIcon(ZulipIcons.user)));
```
Hmm, the use of `Center` is awkward — it's rather an implementation detail, both the fact that the desired copy of the icon does have a Center ancestor and the fact that the undesired copies don't.

I don't have a better idea to suggest, though, so I guess just leave it. No need for a TODO, even; really it just becomes actionable if the test breaks (because of some innocent change to either the nav bar, or how DMs appear in the inbox, or something else) and we have to find a different solution.

I feel like the API I'd want here is to be able to say "tap the such-and-such icon, specifically the one that's in the bottom part of the screen". But I don't see an easy way to do that with the existing `Finder` API.

PIG208 replied:
Yeah, I was a bit reluctant when writing this, but some alternatives would be customizing the initial view, adding a debug feature, or anything that only gets used in tests like this, which are not really realistic/useful. Leaving it seems fine to me.

PIG208 commented on lib/widgets/home.dart:
```diff
@@ -1,97 +1,597 @@
+          mainAxisSize: MainAxisSize.min,
+          children: [
+            const CircularProgressIndicator(),
+            IgnorePointer(
+              ignoring: !showTryAnotherAccount,
+              child: Opacity(opacity: showTryAnotherAccount ? 1 : 0,
+                child: Padding(
```
Needed to change this to:
```dart
            Visibility(
              visible: showTryAnotherAccount,
              maintainSize: true,
              maintainAnimation: true,
              maintainState: true,
```

```
  /// Whether to maintain space for where the widget would have been.
  ///
  /// To set this, [maintainAnimation] and [maintainState] must also be set.
  /// [...]
  final bool maintainSize;
```

gnprice commented on test/widgets/actions_test.dart:
```diff
@@ -182,8 +182,11 @@ void main() {
       testNavObserver.onPushed = (route, prevRoute) => pushedRoutes.add(route);
-      final account1Route = InboxPage.buildRoute(accountId: account1.id);
-      final account2Route = InboxPage.buildRoute(accountId: account2.id);
+      // TODO(#737): switch to a realistic setup:
+      final account1Route = MaterialAccountWidgetRoute(
```
bump https://github.com/zulip/zulip-flutter/pull/1076#discussion_r1879099666 :slightly_smiling_face:

PIG208 replied:
Ohh! I was wondering why there was only a colon added (which made some sense to me anyway).

gnprice commented on lib/widgets/home.dart:
```diff
@@ -1,97 +1,598 @@
+  Widget buildLeading(BuildContext context) {
+    assert(icon != null);
+    final designVariables = DesignVariables.of(context);
+    return Icon(icon, size: _iconSize, color: selected ? designVariables.iconSelected
+                                                : designVariables.icon);
```
nit: misaligned

(probably just move `color` to second line)

gnprice commented on lib/widgets/home.dart:
```diff
@@ -1,97 +1,598 @@
+  Widget buildLeading(BuildContext context) {
+    final store = PerAccountStoreWidget.of(context);
+    return Avatar(
+      userId: store.selfUserId, size: _MenuButton._iconSize, borderRadius: 4);
```
Ah, bonus factoring-out! :slightly_smiling_face:

gnprice reviewed (commented):
Thanks for the swift revision! Just two nits now.

gnprice commented:
@PIG208 would you rebase? Looks like the last commit has a non-local conflict with #1041 (should have merged this one before that one, I guess, oops).

gnprice commented:
Otherwise all LGTM!

PIG208 commented:
Rebased and fixed the conflict with `$ git rebase -x 'tools/check l10n --fix'`. Additionally addressed the issue that `InboxPage` no longer exists for a test from #1041.

gnprice commented:
Thanks!

CI has reached the `android` suite (which is the last one because it's slow), and it looks like the others pass. That one shouldn't be affected by these changes, so merging.
